controllers/effective_dnspolicies_reconciler.go

```diff
@@ -0,0 +1,290 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/samber/lo"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	kuadrantdnsv1alpha1 "github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+)
+
+func NewEffectiveDNSPoliciesReconciler(client *dynamic.DynamicClient, scheme *runtime.Scheme) *EffectiveDNSPoliciesReconciler {
+	return &EffectiveDNSPoliciesReconciler{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+type EffectiveDNSPoliciesReconciler struct {
+	client *dynamic.DynamicClient
+	scheme *runtime.Scheme
+}
+
+func (r *EffectiveDNSPoliciesReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.DNSPolicyGroupKind},
+			{Kind: &DNSRecordGroupKind},
+		},
+	}
+}
+
+func (r *EffectiveDNSPoliciesReconciler) reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("EffectiveDNSPoliciesReconciler")
+
+	policyTypeFilterFunc := dnsPolicyTypeFilterFunc()
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	policyErrors := map[string]error{}
+
+	logger.V(1).Info("updating dns policies", "policies", len(policies))
+
+	for _, policy := range policies {
+		pLogger := logger.WithValues("policy", policy.GetLocator())
+
+		if policy.GetDeletionTimestamp() != nil {
+			pLogger.V(1).Info("policy marked for deletion, skipping")
+			continue
+		}
+
+		if accepted, _ := policyAcceptedFunc(policy); !accepted {
+			pLogger.V(1).Info("policy not accepted, skipping")
+			continue
+		}
+
+		listeners := r.listenersForPolicy(ctx, topology, policy, policyTypeFilterFunc)
+
+		if logger.V(1).Enabled() {
+			listenerLocators := lo.Map(listeners, func(item *machinery.Listener, _ int) string {
+				return item.GetLocator()
+			})
+			pLogger.V(1).Info("reconciling policy for gateway listeners", "listeners", listenerLocators)
+		}
+
+		var gatewayHasAttachedRoutes = false
+		var gatewayHasAddresses = false
+
+		clusterID, err := utils.GetClusterUID(ctx, r.client)
+		if err != nil {
+			return fmt.Errorf("failed to generate cluster ID: %w", err)
+		}
```

Comment:
Can we move this outside the loop since we I'm assuming we only need to get this once overall? 🤔

Reply:
yes we can